### PR TITLE
Render graph lines with less curves

### LINKED ISSUE #5782: Readability revision graph
**Do you want to request a *feature* or report a *bug*?**
Discussion/feature

**What is the current behavior?**
The revision graph is very curly. I think the readability could be improved easily. I propose to make the change as visible in this screenprint (left is new):
image

POC: #5783

**Environment you encounter the issue:**
- Git Extensions 3.00.rc2
- d6496b811db20f8890cb3a50caa14937f3e0b650  (Dirty)
- Git 2.19.1.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3110.0
- DPI X:100,00% Y:100,00%

### PR DESCRIPTION
**Pull request by mstv**
Fixes #5782
Inspired by @spdr870's PoC 

## Proposed changes

- Render graph with diagonal lines
- Straighten multi-lane crossings by turning into diagonals
- Unfold one-lane shifts to diagonals
- Join multi-lane crossings
- Both options are active by default - but only as "Experimental" settings.

### Before / After

image image   image image

image image

## Test methodology 

- used for long time in private builds
- adapted existing test cases
- several new test cases 

## Merge strategy

Please do **not squash** merge this PR!

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
Just looking at the screenshots, and I'm not fond of the new version, sorry.
-1 from me.

**Comment by mstv** [comment]
This a draft PoC. I wanted to know whether it is worth the effort to smooth the kinks or whether I rather should not bother you with this again.

image

**Comment by gerhardol** [comment]
+1 from me 
Have been using this occasionally too.

**Comment by RussKie** [comment]
> I wanted to know whether it is worth the effort to smooth the kinks or whether I rather should not bother you with this again.

We all have likes and dislikes :) And if others like this way better - we'll have it, especially since there's an option to turn it off.

**Comment by mstv** [comment]
> Any progress?

yes, @xnousnow

**Comment by RussKie** [comment]
> ### Before / After
> image image image image

Are these reflecting the latest state?

**Comment by mstv** [comment]
> Are these reflecting the latest state?

Yes, they are. The drawing of the graph is ready for review.
(In a follow-up, I am going to turn multi-line crossings into "diagonals": 13776d6bc7283f6155d6b16d3121ec83dac7df33. This is done in the same lower layer as #10778. I.e. it is an independent change.)
 
Open points (not affecting the graph):
* Avoid often instantiation of `SegmentDrawer` perhaps as in a44f5a8309239b2394b64ca9547f65ac3c0ab2b2? Code smell "static instance" though.
* `AppSettings.ReduceGraphCurves` to be added to a settings page?

File: GitCommands/Settings/AppSettings.cs [hunk 1]
```diff
@@ -1499,6 +1499,12 @@ public static bool FillRefLabels
             set => SetBool("FillRefLabels", value);
         }
 
+        public static bool ReduceGraphCurves
+        {
+            get => GetBool("ReduceGraphCurves", true);
+            set => SetBool("ReduceGraphCurves", value);
+        }
```
Comment (RussKie) [inline GitCommands/Settings/AppSettings.cs:1506]: I'm thinking more and more that we need to have an "Experimental" section in settings where we can expose settings and easily toggle those.

File: GitUI/UserControls/RevisionGrid/Graph/Rendering/DiagonalSegmentInfo.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+﻿namespace GitUI.UserControls.RevisionGrid.Graph.Rendering
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/Graph/Rendering/DiagonalSegmentInfo.cs:1]:
```suggestion
﻿namespace GitUI.UserControls.RevisionGrid.Graph.Rendering;
```

**Review by RussKie — COMMENTED** [review]
:shipit:

**Comment by RussKie** [comment]
We probably should move all graph related settings on their own dedicated page but in a separate PR :)

**Comment by pmiossec** [comment]
@mstv Can the `verify.txt` files be renamed to be shorter?

Because there is for example the file `UnitTests\GitUI.Tests\UserControls\RevisionGrid\Graph\RevisionGraphTests.UnfoldOneLaneShiftsToDiagonals_commitSpecs=0-D,1,2,3,4,5,6,B,8,9,7 1-D 2-D 3-E 4-E 5-C 6-C 7-G 8-9 9-A A-B B-C C-F D-R E-F F-R G-H H-R R.verified.txt` which is 223 characters long and it will be painfull for everyone to handle (and especially the new contributors).

As a consequence, I can't do anything with master due to error:

```
error: cannot stat 'UnitTests/GitUI.Tests/UserControls/RevisionGrid/Graph/RevisionGraphTests.UnfoldOneLaneShiftsToDiagonals_commitSpecs=0-D,1,2,3,4,5,6,B,8,9,7 1-D 2-D 3-E 4-E 5-C 6-C 7-G 8-9 9-A A-B B-C C-F D-R E-F F-R G-H H-R R.verified.txt': Filename too long
```

**Comment by mstv** [comment]
> Can the verify.txt files be renamed to be shorter?

I am the wrong addressee. I have not pursued the switch to it (and I am not 100% convinced of it. E.g. comments cannot be added in place).

> which is 223 characters long and it will be painful for everyone to handle

Which kind of problems other than MSYS Git issues? I have never noticed any - and I have rebased these commits many times and have checked them out on a few machines. Though in my checkout, the pathname does not exceed 260 characters.
Perhaps `git config --global core.longpaths true` helps (refer to https://stackoverflow.com/questions/22575662/filename-too-long-in-git-for-windows)

**Comment by RussKie** [comment]
I can the see the point of renaming this test, it is long and can cause some issues.
Verify tests allow customisation of the test names. I'd go with https://github.com/VerifyTests/Verify/blob/main/docs/naming.md#testclassname, e.g. something like:
```cs
        [Test]
        [TestCase("0:1,4 1:2 2:R,3 3:R 4:R R", "case1")]
        [TestCase("0:1,R 1:2,R 2:4,3 3:R 4:5,8 5:6 6:7,R 7:R,R 8:9 9:R R", "case2")]
        [TestCase("0:D,1,2,3,4,5,6,B,8,9,7 1:D 2:D 3:E 4:E 5:C 6:C 7:G 8:9 9:A A:B B:C C:F D:R E:F F:R G:H H:R R", "case3")]
        [TestCase("0:4 1:3,2 2:6 3:8,B 4:5 5:A,6 6:7 7:8 8:C,R 9:A A:B,R B:R C:R R", "case4")]
        public async Task UnfoldOneLaneShiftsToDiagonals(string commitSpecs, string testNameSuffix)
        {
            AppSettings.StraightenGraphDiagonals.Value = true;

            RevisionGraph revisionGraph = CreateGraphTopDown(commitSpecs);

            await VerifyGraphLayoutAsync(revisionGraph, $"{MethodBase.GetCurrentMethod().Name}_{testNameSuffix}");
        }

        private static async Task VerifyGraphLayoutAsync(RevisionGraph revisionGraph, string testNameSuffix)
        {
            string actualGraph = AsciiGraphFor(revisionGraph).Join("\n");
            await Verify(actualGraph)
                .UseMethodName(testNameSuffix);
        }
```

**Comment by RussKie** [comment]
> E.g. comments cannot be added in place

I'm not quite sure I follow, could you please elaborate?

**Comment by mstv** [comment]
> > E.g. comments cannot be added in place
> 
> I'm not quite sure I follow, could you please elaborate?

Comments like the following would benefit from having the ASCII graph in the code - but not in a dedicated file.

```
        // Do not move the right lane of row 2.
        [TestCase("0:1,5 1:2   2:R,3 3:R,4 4:R   5:R R")]
        // Do not move the right lanes of row 6.
        [TestCase("0:1,R 1:2,R 2:4,3 3:R   4:5,R 5:6   6:7   7:8,R 8:9 9:R R")]
        // Do not move the right lanes of row 4.
        [TestCase("0:1,R 1:2,R 2:4,3 3:R 4:5,8 5:6,R 6:7,R 7:R,R 8:9 9:R R")]
        // Do not move segment 4:8 at row 6 & 5.
        [TestCase("0:5,1,2,3,4 1:5 2:5 3:6 4:8 5:R 6:7 7:8,B 8:9 9:R,A A:R   B:R     R")]
        // Do not move segment 7:B at row 9.
        [TestCase("0:5,1,2,3,4 1:5 2:5 3:6 4:8 5:R 6:7 7:8,B 8:9 9:R,A A:R,C B:R C:R R")]
```

Though since I do not expect many further changes to the graph rendering, this does not matter very much.

**Pull request merged**